File: serverstandardstore/src/main/java/com/lightbend/standard/modelserver/DataProcessor.java
package com.lightbend.standard.modelserver;

import com.lightbend.configuration.kafka.ApplicationKafkaParameters;
import com.lightbend.custom.modelserver.store.ModelStateStore;
import com.lightbend.custom.queriablestate.ModelServingInfo;
import com.lightbend.model.Winerecord;
import org.apache.kafka.streams.processor.AbstractProcessor;
import org.apache.kafka.streams.processor.ProcessorContext;

import java.util.Objects;
import java.util.Optional;

/**
 * Created by boris on 7/12/17.
 * used
 * https://github.com/bbejeck/kafka-streams/blob/master/src/main/java/bbejeck/processor/stocks/StockSummaryProcessor.java
 */
public class DataProcessor extends AbstractProcessor<byte[], Optional<Winerecord.WineRecord>> {

    private ModelStateStore modelStore;

    @Override
    public void process(byte[] key, Optional<Winerecord.WineRecord> dataRecord) {
        if(modelStore.getNewModel() != null){
            // update the model
            if(modelStore.getCurrentModel() != null)
                modelStore.getCurrentModel().cleanup();
            modelStore.setCurrentModel(modelStore.getNewModel());
            modelStore.setCurrentServingInfo(new ModelServingInfo(modelStore.getNewServingInfo().getName(),
                    modelStore.getNewServingInfo().getDescription(), System.currentTimeMillis()));
            modelStore.setNewServingInfo(null);
            modelStore.setNewModel(null);
        }
        // Actually score
        if(modelStore.getCurrentModel() == null) {
            // No model currently
            System.out.println("No model available - skipping");
        }
        else{
            // Score the model
            long start = System.currentTimeMillis();
            double quality = (double) modelStore.getCurrentModel().score(dataRecord.get());
            long duration = System.currentTimeMillis() - start;
            modelStore.getCurrentServingInfo().update(duration);
            System.out.println("Calculated quality - " + quality + " in " + duration + "ms");
         }

    }

    @Override
    public void init(ProcessorContext context) {
        modelStore = (ModelStateStore) context.getStateStore(ApplicationKafkaParameters.STORE_NAME);
        Objects.requireNonNull(modelStore, "State store can't be null");

    }
}

File: kafkaStreamsModelServerCustomStore/src/main/scala/com/lightbend/scala/custom/DataProcessor.scala
package com.lightbend.scala.custom

import java.util.Objects

import com.lightbend.configuration.kafka.ApplicationKafkaParameters
import com.lightbend.model.winerecord.WineRecord
import org.apache.kafka.streams.processor.{AbstractProcessor, ProcessorContext, ProcessorSupplier}
import com.lightbend.scala.custom.store.ModelStateStore

import scala.util.Try

class DataProcessor extends AbstractProcessor[Array[Byte], Try[WineRecord]] with ProcessorSupplier[Array[Byte], Try[WineRecord]]{

  private var modelStore = null.asInstanceOf[ModelStateStore]

  import ApplicationKafkaParameters._

  override def process(key: Array[Byte], dataRecord: Try[WineRecord]): Unit = {

    modelStore.state.newModel match {
      case Some(model) => {
        // close current model first
        modelStore.state.currentModel match {
          case Some(m) => m.cleanup()
          case _ =>
        }
        // Update model
        modelStore.state.currentModel = modelStore.state.newModel
        modelStore.state.currentState = modelStore.state.newState
        modelStore.state.newModel = None
      }
      case _ =>
    }
    modelStore.state.currentModel match {
      case Some(model) => {
        val start = System.currentTimeMillis()
        val quality = model.score(dataRecord.get.asInstanceOf[AnyVal]).asInstanceOf[Double]
        val duration = System.currentTimeMillis() - start
        println(s"Calculated quality - $quality calculated in $duration ms")
        modelStore.state.currentState.get.incrementUsage(duration)
      }
      case _ => {
        println("No model available - skipping")
      }
    }
   }

  override def init(context: ProcessorContext): Unit = {
    modelStore = context.getStateStore(STORE_NAME).asInstanceOf[ModelStateStore];
    Objects.requireNonNull(modelStore, "State store can't be null")
  }

  override def get() = new DataProcessor()
}
